omnigibson/controllers/osc_controller.py

```diff
@@ -0,0 +1,535 @@
+import numpy as np
+from numba import jit
+
+import omnigibson.utils.transform_utils as T
+from omnigibson.controllers import ControlType, ManipulationController
+from omnigibson.utils.control_utils import orientation_error
+from omnigibson.utils.processing_utils import MovingAverageFilter
+from omnigibson.utils.python_utils import nums2array, assert_valid_key
+from omnigibson.utils.ui_utils import create_module_logger
+
+# Create module logger
+log = create_module_logger(module_name=__name__)
+
+# Different modes
+OSC_MODE_COMMAND_DIMS = {
+    "absolute_pose": 6,      # 6DOF (x,y,z,ax,ay,az) control of pose, whether both position and orientation is given in absolute coordinates
+    "pose_absolute_ori": 6,  # 6DOF (dx,dy,dz,ax,ay,az) control over pose, where the orientation is given in absolute axis-angle coordinates
+    "pose_delta_ori": 6,  # 6DOF (dx,dy,dz,dax,day,daz) control over pose
+    "position_fixed_ori": 3,  # 3DOF (dx,dy,dz) control over position, with orientation commands being kept as fixed initial absolute orientation
+    "position_compliant_ori": 3,  # 3DOF (dx,dy,dz) control over position, with orientation commands automatically being sent as 0s (so can drift over time)
+}
+OSC_MODES = set(OSC_MODE_COMMAND_DIMS.keys())
+
+
+class OperationalSpaceController(ManipulationController):
+    """
+    Controller class to convert (delta) EEF commands into joint efforts using Operational Space Control
```

Comment:
why is delta in parenthesis? Can OSC convert absolute EEF commands to joint efforts? or only delta ones?

Reply:
updated